### ISSUE TITLE
"mode()" evaluates to "v" after entering the search command-line from visual mode

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#6127]
**Describe the bug**

`mode()` evaluates to `v` after entering the search command-line from visual mode.

**To Reproduce**

Start Vim with this shell command:

    vim -Nu NONE +'cno <expr> <c-b> mode()'

Press `v` to enter visual mode, then `/` to enter the search command-line, then `C-b` to invoke the mapping which inserts the current mode on the command-line.

`v` is inserted.

**Expected behavior**

`c` is inserted.

**Environment**

 - Vim version: 8.2 Included patches: 1-813
 - OS: Ubuntu 16.04.6 LTS
 - Terminal: XTerm(322)

**Additional context**

Even though the command-line was entered from visual mode, it doesn't make sense for `mode()` to return `v`; especially when the latter is invoked right from a mapping installed in *command-line* mode.

---

If it's a special case, then it's an inconsistent one, because if you try the same experiment, but you enter the command-line with `:` instead of `/`, this time you get the expected `c` mode:

                                                              v
    vim -Nu NONE +'cno <expr> <c-b> mode()' +'call feedkeys("v:\<c-b>")'
    :'<,'>c
          ^

**Comment by lacygoill** [comment]
I forgot one step in the original post to reproduce. You need to press `/` to enter the search command-line. I edited the post.

Here's a one-liner to illustrate the unexpected output of `mode()`:

    vim -Nu NONE +'cno <expr> <c-b> mode()' +'call feedkeys("v/\<c-b>")'

**Comment by brammool** [comment]
Hmm, mode() checks VIsual_active before whether State is CMDLINE.
I hesitate to change this, since it's been like this for a long time.
On the other hand, it does make sense to let the temporary "command line" mode trumpthe Visual state.

**Comment by lacygoill** [comment]
> I hesitate to change this, since it's been like this for a long time.

FWIW, it causes a bug in one of my plugins which assumes `mode()` to be `c` on the command-line, regardless of the mode from which the latter was entered.  I guess I could work around by substituting `v`, `V` and `^V` with `c`, but I wonder whether it has caused me other issues in the past.

I admit I don't know whether fixing it now would do more harm than good in existing scripts...
Maybe `mode()` could still return `v`, but `mode(1)` would at least return something different to let the user know that they're really on the command-line; something like `vc`.

**Comment by tonymec** [comment]
Hm, mode() return value of "cv" already exists for Ex mode. What about "cV" for command-line mode triggered from Visual? Then a plugin could test `if mode() =~ '^c'` to test for command-line mode *lato sensu*, including Ex mode.

Best regards,
Tony.

**Comment by brammool** [comment]
> > I hesitate to change this, since it's been like this for a long time.
> 
> FWIW, it causes a bug in one of my plugins which assumes `mode()` to
> be `c` on the command-line, regardless of the mode from which the
> latter was entered.  I guess I could work around by substituting `v`,
> `V` and `^V` with `c`, but I wonder whether it has caused me other
> issues in the past.
> 
> I admit I don't know whether fixing it now would do more harm than
> good in existing scripts...  Maybe `mode()` could still return `v`,
> but `mode(1)` would at least return something different to let the
> user know that they're really on the command-line; something like
> `vc`.

There are actually more situations where the main mode is command-line,
but when done we end up in another mode.  E.g. "d/" is operator-pending.

Perhaps we can use something like "c/v" for  using "/" in Visual mode
and "c/o" for using "/" after an operator.

-- 
Living in Hollywood is like living in a bowl of granola.  What ain't
fruits and nuts is flakes.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by lacygoill** [comment]
Yes, that sounds good.  We could then write `if mode(1) =~# '^c'` to check if we're on the command-line, even if the latter was entered from visual mode or operator-pending mode; and inspect the rest of `mode(1)` if we need to know the precise stack of modes.

**Comment by LemonBoy** [comment]
> There are actually more situations where the main mode is command-line, but when done we end up in another mode. E.g. "d/" is operator-pending. Perhaps we can use something like "c/v" for using "/" in Visual mode and "c/o" for using "/" after an operator.

If I understand correctly your proposal is to return a set of modes like this:
```
'c' {':', '/', '?', '>', '='} {'v', 'V', 'o'}
 ^  ^- cmdfirstc              ^- sub-mode
 `------ command line mode
```
IMO it's quite heavy as it replicates information that the user can already retrieve using `getcmdtype()` if needed.
Also, documenting all the possible combinations in the help text for `mode()` is going to be quite some (trivial) work.

**Comment by zeertzjq** [comment]
A similar problem can happen when calling `mode()` in a callback at a hit-enter/more/confirm prompt in Visual mode, although that can be checked using `s` flag returned by `state()`.

**Comment by zeertzjq** [comment]
Fixed by patch 9.1.0683
